pkg/target/manager.go

```diff
@@ -0,0 +1,399 @@
+		seed, err := m.resolveSeedName(ctx, gardenClient, seedName)
+		if err != nil {
+			return fmt.Errorf("failed to validate seed: %w", err)
+		}
+
+		t.Seed = seedName
+		t.Project = ""
+		t.Shoot = ""
+
+		if err := m.updateSeedKubeconfig(ctx, gardenClient, t, seed); err != nil {
+			return fmt.Errorf("failed to fetch kubeconfig for seed cluster: %w", err)
+		}
+
+		return nil
+	})
+}
+
+func (m *managerImpl) updateSeedKubeconfig(ctx context.Context, gardenClient client.Client, t Target, seed *gardencorev1beta1.Seed) error {
+	// fetch kubeconfig secret
+	secret := corev1.Secret{}
+	key := types.NamespacedName{Name: seed.Spec.SecretRef.Name, Namespace: seed.Spec.SecretRef.Namespace}
+
+	if err := gardenClient.Get(ctx, key, &secret); err != nil {
+		return fmt.Errorf("failed to retrieve seed kubeconfig: %w", err)
+	}
+
+	return m.kubeconfigCache.Write(t, secret.Data["kubeconfig"])
+}
+
+func (m *managerImpl) resolveSeedName(ctx context.Context, gardenClient client.Client, seedName string) (*gardencorev1beta1.Seed, error) {
+	seed := &gardencorev1beta1.Seed{}
+	key := types.NamespacedName{Name: seedName}
+
+	if err := gardenClient.Get(ctx, key, seed); err != nil {
+		return nil, err
+	}
+
+	if seed.Spec.SecretRef == nil {
+		return nil, errors.New("spec.SecretRef is missing in this seed, seed not reachable")
+	}
+
+	return seed, nil
+}
+
+func (m *managerImpl) TargetShoot(ctx context.Context, shootName string) error {
+	return m.patchTarget(func(t *targetImpl) error {
+		if t.Garden == "" {
+			return ErrNoGardenTargeted
+		}
+
+		if t.Project == "" && t.Seed == "" {
+			return errors.New("must target project or seed first")
+		}
+
+		gardenClient, err := m.clientForGarden(t.Garden)
+		if err != nil {
+			return fmt.Errorf("could not create Kubernetes client for garden cluster: %w", err)
+		}
+
+		var (
+			project *gardencorev1beta1.Project
+			seed    *gardencorev1beta1.Seed
+		)
+
+		if t.Project != "" {
+			project, err = m.resolveProjectName(ctx, gardenClient, t.Project)
+			if err != nil {
+				return fmt.Errorf("failed to validate project: %w", err)
+			}
+		} else {
+			seed, err = m.resolveSeedName(ctx, gardenClient, t.Seed)
+			if err != nil {
+				return fmt.Errorf("failed to fetch kubeconfig for seed cluster: %w", err)
+			}
+		}
+
+		t.Shoot = shootName
+
+		if err := m.updateShootKubeconfig(ctx, gardenClient, project, seed, t); err != nil {
+			return fmt.Errorf("failed to fetch kubeconfig for shoot cluster: %w", err)
+		}
+
+		return nil
+	})
+}
+
+// updateShootKubeconfig must be called with *either* project or seed, but never both
+// and never neither.
+func (m *managerImpl) updateShootKubeconfig(
+	ctx context.Context,
+	gardenClient client.Client,
+	project *gardencorev1beta1.Project,
+	seed *gardencorev1beta1.Seed,
+	t Target,
+) error {
+	shoot := &gardencorev1beta1.Shoot{}
+
+	// If a shoot is targeted via a project, we fetch the project and find
+	// the shoot by listing all shoots in the project's spec.namespace.
```

Comment:
This comment is not correct. If we have the project / namespace, we can just get the shoot (as you do below). No need to list anything